ic-task-scheduler/src/scheduler.rs

```diff
@@ -1,86 +1,170 @@
+use std::future::Future;
+use std::pin::Pin;
 use std::sync::Arc;
 
-use ic_stable_structures::UnboundedMapStructure;
+use ic_kit::RejectionCode;
+use ic_stable_structures::{BTreeMapStructure as _, HeapBTreeMap, IterableUnboundedMapStructure};
 use parking_lot::Mutex;
 
 use crate::task::{ScheduledTask, Task};
 use crate::time::time_secs;
-use crate::SchedulerError;
+use crate::{Result, SchedulerError};
+
+/// Internal type used to store tasks to be processed and processing tasks
+type TaskQueue = Arc<Mutex<HeapBTreeMap<u32, (), ()>>>;
+
+/// The state of a task execution.
+/// This is reported when the `SaveStateQueryCallback` is called.
+#[derive(Debug)]
+pub enum TaskExecutionState {
+    /// Reported when tasks to be executed are scheduled
+    Scheduled,
+    /// Reported when a task fails.
+    Failed(u32, SchedulerError),
+    /// Reported when a task starts executing.
+    Executing(u32),
+    /// Reported when a task completes successfully.
+    Completed(u32),
+    /// Reported when a task panics.
+    Panicked(u32),
+}
 
-type SchedulerErrorCallback<T> = Box<dyn 'static + Fn(ScheduledTask<T>, SchedulerError) + Send>;
+type SaveStateQueryCallback = dyn Fn(
+        TaskExecutionState,
+    ) -> Pin<Box<dyn Future<Output = std::result::Result<(), (RejectionCode, String)>>>>
```

Comment:
why are you forcing a RejectionCode as result? it should return a scheduler error if needed